OAuth: Add setting to skip org assignment for external users

prskr opened the pull request:
**What this PR does / why we need it**:

Introduce `skip_social_org_role_assign` setting to skip any kind of org assignment during the login of external users.
As a consequence manual organization assignments won't be overridden during the upsert of an external user.

**Which issue(s) this PR fixes**:

Fixes #22605

**Special notes for your reviewer**:

achatterjee-grafana commented on docs/sources/administration/configuration.md:
```diff
@@ -623,6 +623,13 @@ options are `Admin` and `Editor`. e.g.:
 
+### skip_social_org_role_assign
+
+Skip forced assignment of OrgID `1` or `auto_assign_org_id` for social logins.
+This allows to manually distribute users with social login to multiple organizations.
+Otherwise, the users' organization would be reset on every new login e.g. via AzureAD.
+Default is `false`.
```
```suggestion
Skip forced assignment of OrgID `1` or `auto_assign_org_id` for social logins. Default is `false`.
Use this setting to distribute users with a social login to multiple organizations.
Otherwise, the users' organization would get reset on every new login, for example, via AzureAD.
```

baez90 replied:
thank you very much for the feedback!
I replaced my original draft with your suggestion :blush:

achatterjee-grafana reviewed (commented):
Added doc change request.

baez90 commented:
@marefr I'd be happy to discuss everything you like/dislike about this idea :blush: 

Let me please explain why I think this is at least an 'okay-ish' option:

The original intention was to allow any manual assignment of organizations instead of enforcing a single organization.

Having in mind that teams are favored over organizations and implemented a sync of organizations e.g. based on OIDC claims would require a much bigger effort I thought at least allowing manual assignments would be a good compromise to still keep teams as the more powerful mechanism in place with auto-sync and so on but allow to use multiple organizations in combination with social logins in any way.

torkelo commented:
Not sure about this, or at least the naming feels off (Social, but will be applied to most external auth). Maybe SkipOrgRoleUpdateSync ? 

But the best way would be to allow some kind of override flag via the UI when setting the role.

baez90 commented:
I'm all in for a better naming! I'm not really happy either with the current name but wasn't sure how conventions are and what it really affects and so on and so forth.

Probably `skip_org_role_sync` or `skip_org_role_update` ?
`update_sync` feels like kinda doubled but can also name `skip_org_role_update_sync` if you prefer!

Regarding the idea of the UI switch: you mean something like a `StaticallyAssigned bool` field in the `OrgUser` model that can be set in the UI and that skips the removal of the `OrgUser` in the `loginservice.go`?

While I'm relatively sure I could do the Go stuff I might need some assistance with the UI stuff but if that's okay I could give it a shot

prskr committed:
Chore: add setting to skip org assignment for external users

Introduce 'skip_org_role_update_sync' setting to skip any kind of org assignment during the login of external users.
As a consequence manual organization assignments won't be overridden during the upsert of an external user.

Part of #22605

baez90 commented:
Renamed it to `SkipOrgRoleUpdateSync` like proposed

baez90 commented:
Is there anything missing, anything to discuss or anything else I can do to get some feedback/progress on this or shall I just close it as it won't be merged anyway?

baez90 commented:
@torkelo any chance for some feedback?

I also thought about a follow up PR to add a webhook functionality to notify external systems about newly logged in users.

This would really help us to integrate Grafana with our rather complex permission management but I'm not sure if this aligns with your overall strategy?

achatterjee-grafana reviewed (commented):
Added doc review.

achatterjee-grafana reviewed (approved):
Doc content looks good.

igorhrq commented:
When this will release, probably on next release?

igorhrq commented:
@achatterjee-grafana any news here please? Really appreciate if you guy work on this PR

joningib commented:
Likewise, I would also greatly appreciate this feature being implemented soon :)

joningib commented:
@achatterjee-grafana is there any news on when this might be merged and included in an upcoming version update? Thank you for your contribution on this feature improvement!

achatterjee-grafana commented:
Hello @torkelo and @marefr, what should be the steps forward? Thank you for any clarification.

marefr commented on conf/defaults.ini:
```diff
@@ -333,6 +333,9 @@ auto_assign_org_id = 1
 # Default role new users will be automatically assigned (if auto_assign_org above is set to true)
 auto_assign_org_role = Viewer
 
+# Skip forced assignment of OrgID 1 or 'auto_assign_org_id'  for social logins
+skip_org_role_update_sync = false
+
```
First, sample.ini needs the same update, but out-commented. 

Second, since this is a setting only applied for OAuth feels like the setting should be under the section `[auth]`, rather than under `[users]`, and be named `oauth_skip_org_role_update_sync`. With these changes need to update the documentation/configuration.md as well.

Thoughts?

baez90 replied:
Sounds reasonable, I'll apply this tomorrow if no one else has a different opinion?
Not sure if `oauth` or `social` fits better as a prefix but to be honest, I've neither a strong opinion nor a 'good feeling' about code style and structure in Grafana hence I'd be glad to follow your lead.

baez90 replied:
I renamed and relocated the setting as requested. For now I pushed it as a second commit but I can squash them as soon as the final location is found

baez90 replied:
@achatterjee-grafana I also updated the docs and moved the corresponding part to the `auth` section of course. Apart from that I didn't change anything besides the header obviously

marefr reviewed (commented):
@baez90 thanks. Sorry for the slow response. I think this looks acceptable besides a few things. I'll need to ask some additional people to have a look at this as well and I'll get back soon. Until then please have a look at my comments.

stefanpetter commented:
@sakjur @alexanderzobnin @pkolyvas @osg-grafana Apoligies for getting your attention in this way, but it would be really nice if this simple patch can find it's way to Grafana asap. Lack of this feature stalls implementation of Grafana in a multi-tenancy way for a lot of users.

osg-grafana commented:
> @sakjur @alexanderzobnin @pkolyvas @osg-grafana Apoligies for getting your attention in this way, but it would be really nice if this simple patch can find it's way to Grafana asap. Lack of this feature stalls implementation of Grafana in a multi-tenancy way for a lot of users.

Deferring to @achatterjee-grafana for docs.

sakjur commented:
:wave: I've been away for a couple of weeks, thanks for the ping @stefanpetter to bring this to my attention.

Based on a quick look, the feature makes sense, and the implementation is mostly neat (I want to investigate how hard it would be for us to not have to introduce a new global and simply just use the `Cfg` object and whether there are other places to [also?] apply the restrictions).

baez90 commented:
Wouldn't mean a lot of effort, I'd have to attach the `buildExternalUserInfo` function to the `HTTPServer` struct and could access the `Cfg` directly.

Could do that if you prefer

prskr committed:
Chore: Rename skip_org_role_update_sync to oauth_skip_org_role_update_sync and relocate it to auth section

sakjur commented:
@baez90 That would be great, we're trying to reduce the amount of globals with the hope of eventually being able to parallelize tests 😅

prskr committed:
Chore: replace global setting access where possible

baez90 commented:
@sakjur I replaced all global setting accessors where applicable (there's no equivalent for `setting.AppSubUrl` hence I kept that one as it is, to not blow the PR more than already).
Anything else I shall refine? :blush:

sakjur reviewed (commented):
🙏 Thanks, I think this looks really good.

I'll try it out and ask @achatterjee-grafana to take a look at the docs update 🙂

baez90 commented on pkg/setting/setting.go:
```diff
@@ -1252,6 +1254,8 @@ func readAuthSettings(iniFile *ini.File, cfg *Cfg) (err error) {
 	cfg.OAuthCookieMaxAge = auth.Key("oauth_state_cookie_max_age").MustInt(600)
 	SignoutRedirectUrl = valueAsString(auth, "signout_redirect_url", "")
+	cfg.OAuthSkipOrgRoleUpdateSync = auth.Key("oauth_skip_org_role_update_sync").MustBool(false)
+	OAuthSkipOrgRoleUpdateSync = cfg.OAuthSkipOrgRoleUpdateSync
```
Right! I removed it with the last amended commit :blush:

sakjur commented:
The linter complains about goimports for `setting.go`.

```
pkg/setting/setting.go:104: File is not `goimports`-ed (goimports)
```

I'm targeting this for an 8.4.2 release, it seems small and safe, you've waited a long time to get this in, and since 8.4.0 came out yesterday it's going to be a while until 8.5.0 comes around. In case anyone looks to this in the future, that's exceptional and this shouldn't be used as an argument for adding features to patch releases 🙈

stefanpetter commented:
@baez90 You're a legend! Thank you so much :)

sakjur commented:
Manual backport in #45692.
